File: extensions/libdepwin/nappgui/src/osgui/osx/osslider.m
/*
 * NAppGUI Cross-platform C SDK
 * 2015-2024 Francisco Garcia Collado
 * MIT Licence
 * https://nappgui.com/en/legal/license.html
 *
 * File: osslider.m
 *
 */

/* Operating System native slider */

#include "osslider.h"
#include "osslider.inl"
#include "osslider_osx.inl"
#include "oscontrol_osx.inl"
#include "ospanel_osx.inl"
#include "oswindow_osx.inl"
#include "osgui.inl"
#include <core/event.h>
#include <core/heap.h>
#include <sewer/cassert.h>

#if !defined(__MACOS__)
#error This file is only for OSX
#endif

/*---------------------------------------------------------------------------*/

@interface OSXSlider : NSSlider
{
  @public
    uint32_t flags;
    Listener *OnMoved;
}
@end

/*---------------------------------------------------------------------------*/

@implementation OSXSlider

/*---------------------------------------------------------------------------*/

/*
 -(void)drawRect:(NSRect)rect
 {
     NSFrameRect(rect);
     [super drawRect:rect];
 }*/

/*---------------------------------------------------------------------------*/

- (IBAction)onSliderMoved:(id)sender
{
    cassert(sender == self);
    if ([self isEnabled] == YES && self->OnMoved != NULL)
    {
        EvSlider params;
        params.pos = (real32_t)[sender floatValue];
        params.incr = 0;
        params.step = UINT32_MAX;
        listener_event(self->OnMoved, ekGUI_EVENT_SLIDER, (OSSlider *)sender, &params, NULL, OSSlider, EvSlider, void);
    }
}

/*---------------------------------------------------------------------------*/

- (BOOL)acceptsFirstResponder
{
    return YES;
}

/*---------------------------------------------------------------------------*/

- (BOOL)becomeFirstResponder
{
    return YES;
}

/*---------------------------------------------------------------------------*/

- (void)mouseDown:(NSEvent *)theEvent
{
    if (_oswindow_mouse_down((OSControl *)self) == TRUE)
        [super mouseDown:theEvent];
}

@end

/*---------------------------------------------------------------------------*/

/*
static CGFloat i_slider_with_tickmarks_height(const gui_size_t knob_size)
{
    switch(knob_size)
    {
        case ekGUI_MINI:
            return 16.f;
        case ekREGULAR:
            return 18.f;
        case ekGUI_SIZE_LARGE:
            return 26.f;
        cassert_default();
    }

    return 0.f;
}*/

/*---------------------------------------------------------------------------*/

OSSlider *osslider_create(const uint32_t flags)
{
    OSXSlider *slider = nil;
    heap_auditor_add("OSXSlider");
    slider = [[OSXSlider alloc] initWithFrame:NSZeroRect];
    slider->flags = flags;
    slider->OnMoved = NULL;
    _oscontrol_init(slider);
    [slider setMinValue:0.];
    [slider setMaxValue:1.];
    [slider setFloatValue:slider_get_type(flags) == ekSLIDER_HORZ ? 0.f : 1.f];

#if defined(MAC_OS_X_VERSION_10_12) && MAC_OS_X_VERSION_MIN_REQUIRED >= MAC_OS_X_VERSION_10_12
    [slider setVertical:(BOOL)(slider_get_type(flags) == ekSLIDER_VERT)];
#endif

    [slider setAltIncrementValue:0.01];
    [slider setTarget:slider];
    [slider setAction:@selector(onSliderMoved:)];
    [slider setNumberOfTickMarks:0];
    return (OSSlider *)slider;
}

/*---------------------------------------------------------------------------*/

void osslider_destroy(OSSlider **slider)
{
    cassert_no_null(slider);
    cassert_no_null(*slider);
    listener_destroy(&((OSXSlider *)*slider)->OnMoved);
    [*(OSXSlider **)slider release];
    heap_auditor_delete("OSXSlider");
    *slider = NULL;
}

/*---------------------------------------------------------------------------*/

void osslider_OnMoved(OSSlider *slider, Listener *listener)
{
    cassert_no_null(slider);
    cassert_no_null(listener);
    listener_update(&((OSXSlider *)slider)->OnMoved, listener);
}

/*---------------------------------------------------------------------------*/

void osslider_tooltip(OSSlider *slider, const char_t *text)
{
    _oscontrol_tooltip_set((OSXSlider *)slider, text);
}

/*---------------------------------------------------------------------------*/

void osslider_tickmarks(OSSlider *slider, const uint32_t num_tickmarks, const bool_t tickmarks_at_left_top)
{
    cassert_no_null(slider);
    cassert(num_tickmarks > 0);
    unref(slider);
    unref(num_tickmarks);
    unref(tickmarks_at_left_top);
    /*
    switch (((OSXSlider*)slider)->hv)
    {
        case ekGUI_VERTICAL:

            size.width = i_slider_with_tickmarks_height(((OSXSlider*)slider)->knob_size);
            size.height = (CGFloat)((OSXSlider*)slider)->length;
            break;

        case ekGUI_HORIZONTAL:

            size.width = (CGFloat)((OSXSlider*)slider)->length;
            size.height = i_slider_with_tickmarks_height(((OSXSlider*)slider)->knob_size);
            break;

        cassert_default();
    }

    [(OSXSlider*)slider setFrameSize:size];
*/
    /*
   [(OSXSlider*)slider setNumberOfTickMarks:(NSInteger)num_tickmarks];
   [(OSXSlider*)slider setAllowsTickMarkValuesOnly:NO];

#if defined (MAC_OS_X_VERSION_10_12) && MAC_OS_X_VERSION_MIN_REQUIRED >= MAC_OS_X_VERSION_10_12
    [(OSXSlider*)slider setTickMarkPosition:(tickmarks_at_left_top == TRUE) ? NSTickMarkPositionAbove : NSTickMarkPositionBelow];
#else
    [(OSXSlider*)slider setTickMarkPosition:(tickmarks_at_left_top == TRUE) ? NSTickMarkAbove : NSTickMarkBelow];
#endif
*/
}

/*---------------------------------------------------------------------------*/

void osslider_position(OSSlider *slider, const real32_t position)
{
    OSXSlider *lslider = (OSXSlider *)slider;
    real32_t pos = position;
    cassert_no_null(lslider);
    cassert(position >= 0.f && position <= 1.f);
    if (slider_get_type(lslider->flags) == ekSLIDER_VERT)
        pos = 1.f - position;
    [lslider setFloatValue:pos];
}

/*---------------------------------------------------------------------------*/

real32_t osslider_get_position(const OSSlider *slider)
{
    cassert_no_null(slider);
    return (real32_t)[(OSXSlider *)slider floatValue];
}

/*---------------------------------------------------------------------------*/

static real32_t i_slider_without_tickmarks_height(const gui_size_t knob_size)
{
    switch (knob_size)
    {
    case ekGUI_SIZE_MINI:
        return 12.f;
    case ekGUI_SIZE_SMALL:
        return 15.f;
    case ekGUI_SIZE_REGULAR:
        return 25.f;
        cassert_default();
    }

    return 0.f;
}

/*---------------------------------------------------------------------------*/

void osslider_bounds(const OSSlider *slider, const real32_t length, const gui_size_t knob_size, real32_t *width, real32_t *height)
{
    OSXSlider *lslider = (OSXSlider *)slider;
    cassert_no_null(lslider);
    cassert_no_null(width);
    cassert_no_null(height);
    switch (slider_get_type(lslider->flags))
    {
    case ekSLIDER_HORZ:
        *width = length;
        *height = i_slider_without_tickmarks_height(knob_size);
        break;
    case ekSLIDER_VERT:
        *width = i_slider_without_tickmarks_height(knob_size);
        *height = length;
        break;
        cassert_default();
    }
}

/*---------------------------------------------------------------------------*/

void osslider_attach(OSSlider *slider, OSPanel *panel)
{
    _ospanel_attach_control(panel, (NSView *)slider);
}

/*---------------------------------------------------------------------------*/

void osslider_detach(OSSlider *slider, OSPanel *panel)
{
    _ospanel_detach_control(panel, (NSView *)slider);
}

/*---------------------------------------------------------------------------*/

void osslider_visible(OSSlider *slider, const bool_t is_visible)
{
    _oscontrol_set_visible((NSView *)slider, is_visible);
}

/*---------------------------------------------------------------------------*/

void osslider_enabled(OSSlider *slider, const bool_t is_enabled)
{
    _oscontrol_set_enabled((NSControl *)slider, is_enabled);
}

/*---------------------------------------------------------------------------*/

void osslider_size(const OSSlider *slider, real32_t *width, real32_t *height)
{
    _oscontrol_get_size((NSView *)slider, width, height);
}

/*---------------------------------------------------------------------------*/

void osslider_origin(const OSSlider *slider, real32_t *x, real32_t *y)
{
    _oscontrol_get_origin((NSView *)slider, x, y);
}

/*---------------------------------------------------------------------------*/

void osslider_frame(OSSlider *slider, const real32_t x, const real32_t y, const real32_t width, const real32_t height)
{
    _oscontrol_set_frame((NSView *)slider, x, y, width, height);
}

/*---------------------------------------------------------------------------*/

BOOL _osslider_is(NSView *view)
{
    return [view isKindOfClass:[OSXSlider class]];
}

File: extensions/libdepwin/nappgui/src/osgui/gtk3/osslider.c
/*
 * NAppGUI Cross-platform C SDK
 * 2015-2024 Francisco Garcia Collado
 * MIT Licence
 * https://nappgui.com/en/legal/license.html
 *
 * File: osslider.c
 *
 */

/* Operating System native slider */

#include "osslider.h"
#include "osslider.inl"
#include "osgui.inl"
#include "osgui_gtk.inl"
#include "oscontrol_gtk.inl"
#include "ospanel_gtk.inl"
#include "oswindow_gtk.inl"
#include <core/event.h>
#include <core/heap.h>
#include <sewer/cassert.h>

#if !defined(__GTK3__)
#error This file is only for GTK Toolkit
#endif

struct _osslider_t
{
    OSControl control;
    gui_size_t size;
    bool_t launch_event;
    bool_t discrete;
    uint32_t flags;
    Listener *OnMoved;
};

/*---------------------------------------------------------------------------*/

static gboolean i_OnMoved(GtkRange *range, GtkScrollType step, double value, OSSlider *slider)
{
    cassert_unref(slider->control.widget == GTK_WIDGET(range), range);
    unref(step);

    if (slider->launch_event == TRUE && slider->OnMoved != NULL)
    {
        EvSlider params;
        params.pos = value;
        params.incr = 0.f;
        params.step = UINT32_MAX;

        if (value < 0)
            params.pos = 0;
        else if (value > 1)
            params.pos = 1;

        listener_event(slider->OnMoved, ekGUI_EVENT_SLIDER, slider, &params, NULL, OSSlider, EvSlider, void);
        return slider->discrete;
    }

    return FALSE;
}

/*---------------------------------------------------------------------------*/

static gboolean i_OnPressed(GtkWidget *widget, GdkEvent *event, OSSlider *slider)
{
    cassert_no_null(slider);
    unref(widget);
    unref(event);
    if (_oswindow_mouse_down(OSControlPtr(slider)) == TRUE)
        return FALSE;
    return TRUE;
}

/*---------------------------------------------------------------------------*/

OSSlider *osslider_create(const uint32_t flags)
{
    OSSlider *slider = heap_new0(OSSlider);
    GtkOrientation o = slider_get_type(flags) == ekSLIDER_HORZ ? GTK_ORIENTATION_HORIZONTAL : GTK_ORIENTATION_VERTICAL;
    GtkWidget *widget = gtk_scale_new_with_range(o, 0., 1., 0.1);
    gtk_scale_set_draw_value(GTK_SCALE(widget), FALSE);
    gtk_range_set_increments(GTK_RANGE(widget), .1, .1);
    g_signal_connect(widget, "button-press-event", G_CALLBACK(i_OnPressed), (gpointer)slider);
    g_signal_connect(widget, "change-value", G_CALLBACK(i_OnMoved), (gpointer)slider);

    /*
    if (o == GTK_ORIENTATION_HORIZONTAL)
       _oscontrol_widget_set_css(widget, "scale {min-height:5px;padding-left:5px;padding-right:5px}");
    else
       _oscontrol_widget_set_css(widget, "scale {min-height:5px;padding-top:5px;padding-bottom:5px}");
    */

    _oscontrol_init(&slider->control, ekGUI_TYPE_SLIDER, widget, widget, TRUE);
    slider->size = ENUM_MAX(gui_size_t);
    slider->flags = flags;
    slider->launch_event = TRUE;
    slider->discrete = FALSE;
    return slider;
}

/*---------------------------------------------------------------------------*/

void osslider_destroy(OSSlider **slider)
{
    cassert_no_null(slider);
    cassert_no_null(*slider);
    listener_destroy(&(*slider)->OnMoved);
    _oscontrol_destroy(*(OSControl **)slider);
    heap_delete(slider, OSSlider);
}

/*---------------------------------------------------------------------------*/

void osslider_OnMoved(OSSlider *slider, Listener *listener)
{
    cassert_no_null(slider);
    listener_update(&slider->OnMoved, listener);
}

/*---------------------------------------------------------------------------*/

void osslider_tooltip(OSSlider *slider, const char_t *text)
{
    cassert_no_null(slider);
    gtk_widget_set_tooltip_text(slider->control.widget, (const gchar *)text);
}

/*---------------------------------------------------------------------------*/

void osslider_tickmarks(OSSlider *slider, const uint32_t num_tickmarks, const bool_t tickmarks_at_left_top)
{
    gdouble steps = 0.1;
    cassert_no_null(slider);
    cassert(num_tickmarks > 1);

    if (num_tickmarks != UINT32_MAX)
    {
        steps = 1. / (num_tickmarks - 1.);
        slider->discrete = TRUE;
    }
    else
    {
        slider->discrete = FALSE;
    }

    gtk_range_set_increments(GTK_RANGE(slider->control.widget), steps, steps);
    unref(tickmarks_at_left_top);
}

/*---------------------------------------------------------------------------*/

void osslider_position(OSSlider *slider, const real32_t position)
{
    cassert_no_null(slider);
    slider->launch_event = FALSE;
    gtk_range_set_value(GTK_RANGE(slider->control.widget), (gdouble)position);
    slider->launch_event = TRUE;
}

/*---------------------------------------------------------------------------*/

real32_t osslider_get_position(const OSSlider *slider)
{
    cassert_no_null(slider);
    return (real32_t)gtk_range_get_value(GTK_RANGE(slider->control.widget));
}

/*---------------------------------------------------------------------------*/

void osslider_bounds(const OSSlider *slider, const real32_t length, const gui_size_t knob_size, real32_t *width, real32_t *height)
{
    GtkRequisition s;
    cassert_no_null(slider);
    cassert_no_null(width);
    cassert_no_null(height);
    cassert_unref(knob_size == ekGUI_SIZE_REGULAR, knob_size);
    gtk_widget_get_preferred_size(slider->control.widget, &s, NULL);

    switch (slider_get_type(slider->flags))
    {
    case ekSLIDER_HORZ:
        *width = length;
        *height = (real32_t)s.height;
        break;
    case ekSLIDER_VERT:
        *width = (real32_t)s.width;
        *height = length;
        break;
        cassert_default();
    }
}

/*---------------------------------------------------------------------------*/

void osslider_attach(OSSlider *slider, OSPanel *panel)
{
    _ospanel_attach_control(panel, (OSControl *)slider);
}

/*---------------------------------------------------------------------------*/

void osslider_detach(OSSlider *slider, OSPanel *panel)
{
    _ospanel_detach_control(panel, (OSControl *)slider);
}

/*---------------------------------------------------------------------------*/

void osslider_visible(OSSlider *slider, const bool_t is_visible)
{
    _oscontrol_set_visible((OSControl *)slider, is_visible);
}

/*---------------------------------------------------------------------------*/

void osslider_enabled(OSSlider *slider, const bool_t is_enabled)
{
    _oscontrol_set_enabled((OSControl *)slider, is_enabled);
}

/*---------------------------------------------------------------------------*/

void osslider_size(const OSSlider *slider, real32_t *width, real32_t *height)
{
    _oscontrol_get_size((const OSControl *)slider, width, height);
}

/*---------------------------------------------------------------------------*/

void osslider_origin(const OSSlider *slider, real32_t *x, real32_t *y)
{
    _oscontrol_get_origin((const OSControl *)slider, x, y);
}

/*---------------------------------------------------------------------------*/

void osslider_frame(OSSlider *slider, const real32_t x, const real32_t y, const real32_t width, const real32_t height)
{
    _oscontrol_set_frame((OSControl *)slider, x, y, width, height);
}
